File: 0017_Letter Combinations of a Phone Number/LetterCombinationsofaPhoneNumberDFS.java
class Solution {
    public List<String> letterCombinations(String digits) {
        List<String> ans = new ArrayList<>();
        if (digits == null || digits.length() == 0)
            return ans;

        String[] dict = new String[] { " ", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz" };
        char[] cur = new char[digits.length()];
        DFS(digits, dict, 0, cur, ans);
        return ans;
    }

    private void DFS(String digits, String[] dict, int index, char[] cur, List<String> ans) {
        if (index == digits.length()) {
            ans.add(new String(cur));
            return;
        }

        String s = dict[Character.getNumericValue(digits.charAt(index))];

        for (int i = 0; i < s.length(); ++i) {
            cur[index] = s.charAt(i);
            DFS(digits, dict, index + 1, cur, ans);
        }
    }
}

File: 0017_Letter Combinations of a Phone Number/LetterCombinationsofaPhoneNumberDFS.cs
public class Solution {
    private Dictionary<char, List<char>> dict = new Dictionary<char, List<char>>()    
    {
        {'2', new List<char>(){'a','b','c'}},
        {'3', new List<char>(){'d','e','f'}},
        {'4', new List<char>(){'g','h','i'}},
        {'5', new List<char>(){'j','k','l'}},
        {'6', new List<char>(){'m','n','o'}},
        {'7', new List<char>(){'p','q','r','s'}},
        {'8', new List<char>(){'t','u','v'}},
        {'9', new List<char>(){'w','x','y','z'}},
    };
    public IList<string> LetterCombinations(string digits) {
        if(string.IsNullOrEmpty(digits)) return new List<string>();
        
        var ans = new List<string>();
        var list = new char[digits.Length];
        DFS(digits, 0, list, ans, dict);
        
        return ans;
    }
    
    private void DFS(string digits, int index, char[] list, IList<string> ans, Dictionary<char, List<char>> dict)
    {
        if(index == digits.Length)
        {
            ans.Add(new string(list));
            return;
        }
        
        foreach(var c in dict[digits[index]])
        {
            var original = list[index];
            list[index] = c;
            DFS(digits, index+1, list, ans, dict);
            list[index] = original;
            
        }
    }
}